python/mead/downloader.py

```diff
@@ -0,0 +1,300 @@
+import sys
+
+def get_file(file_name):
+    import shutil
+    import os
+    if os.path.exists(file_name):
+        shutil.rmtree(file_name)
+    return file_name
+
+
+def extract_gzip(file_loc):
+    import gzip
+    import shutil
+    temp_file = get_file("{}.1".format(file_loc))
+    with gzip.open(file_loc, 'rb') as f_in:
+        with open(temp_file, 'wb') as f_out:
+            shutil.copyfileobj(f_in, f_out)
+    if mime_type(temp_file) == "application/x-tar":
+        return extract_tar(temp_file)
+    else:
+        shutil.move(temp_file, file_loc)
+        return file_loc
+
+
+def extract_tar(file_loc):
+    import tarfile
+    import os
+    temp_file = get_file("{}.1".format(file_loc))
+    with tarfile.open(file_loc, "r") as tar_ref:
+        tar_ref.extractall(temp_file)
+    if len(os.listdir(temp_file)) != 1:
+        raise RuntimeError("tar extraction unsuccessful")
+    return os.path.join(temp_file, os.listdir(temp_file)[0])
+
+
+def extract_zip(file_loc):
+    import zipfile
+    temp_file = get_file("{}.1".format(file_loc))
+    with zipfile.ZipFile(file_loc, "r") as zip_ref:
+        zip_ref.extractall(temp_file)
+    return temp_file
+
+
+def mime_type(loc):
+    import magic
+    return magic.Magic(mime=True).from_file(loc)
+
+
+def get_cache_dir():
+    import json
+    try:
+        data_cache = json.load(open("config.json")).get("datacache", "~/.bl-dataset-embeddings/")
+    except IOError:
+        data_cache = "~/.bl-dataset-embeddings/"
+
+    return data_cache
+
+
+def extractor(filepath, cache_dir, extractor_func):
+    import hashlib
+    import shutil
+    import os
+    sha1 = hashlib.sha1(open(filepath, 'rb').read()).hexdigest()
+    print("extracting file..")
+    path_to_save = filepath if extractor_func is None else extractor_func(filepath)
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir)
+    path_to_save_sha1 = os.path.join(cache_dir, sha1)
+    shutil.move(path_to_save, path_to_save_sha1)
+    print("downloaded data saved in {}".format(path_to_save_sha1))
+    return path_to_save_sha1
+
+
+def web_downloader(url):
+    import requests
+    import shutil
+    from baseline.progress import create_progress_bar
+    import os
+    r = requests.get(url, stream=True)
+    if r.status_code != 200:
+        raise RuntimeError("The file can not be downloaded")
+    path_to_save = "/tmp/data.dload-{}".format(os.getpid())
+    try:
+        print("downloading {}".format(url))
+        total_length = None
+        with open(path_to_save, 'wb') as f:
+            total_length = r.headers.get('content-length', None)
+            if total_length is not None:
+                total_length = int(total_length)
+                chunk_size = 1024
```

Comment:
Seems like this chunk size is fairly small.  I would expect that larger chunks would be significantly faster